tests/utils.py

```diff
@@ -70,3 +71,24 @@ def expect_failure_if_no_gpu(test):
         return unittest.expectedFailure(test)
     else:
         return test
+
+
+def query_memory(n=2):
+    """
+    find best n idle device ids and return a string of ids
+    """
+    bash_string = "nvidia-smi --query-gpu=utilization.gpu,temperature.gpu,memory.used --format=csv,noheader,nounits"
```

Comment:
@wyli I was really unhappy with the bash script solution.  It just does not feel right to do this.

I wrote a python-based GPU query tool.

[cache_dataset_speed.py.zip](https://github.com/Project-MONAI/MONAI/files/4664617/cache_dataset_speed.py.zip)

Reply:
could you please clarify? simply claiming 'unhappy' is subjective and not very helpful here

Reply:
@wyli Sorry.  that is probably not a good word to use.  I just mean that I would prefer a pure python solution rather than a subprocess call.   I think that the current solution depends on a Unix based environment with bash capabilities. 

It might be fine for the CI environment, but I need similar capabilities that will be cross-platform (i.e. non-Unix dependant) solutions for similar behaviors in my primary tools.

The attached python class is my initial attempt to provide an encapsulated class that allows querying the hardware and returning information about the GPU.

Reply:
@wyli FEEL FREE TO IGNORE MY RECOMMENDATION.  I do not intend to be a roadblock here,  and I am not opposed to your solution.  I only want to provide a potential mechanism that may be more generally used for other purposes.

Reply:
thanks for the explanation @hjmjohnson, I'll try to incorporate your code in another iteration (please feel free to submit a separate PR if you like). My first impression is that using pynvml is very nice, a minor issue is -- ideally this utility should be a self-contained module based on pynvml rather than depending on `torch.cuda` commands.

Reply:
@wyli that is definitely possible.  Originally It expected torch would have all these capabilities, but when I found that it does not provide access to the dynamic elements I found pynvml.

I'll work on fine tuninging my implementation over the next week or so and make a PR.